### ISSUE TITLE
os-ddclient - Ability to choose IPv4 or IPv6 or IPv4/IPv6 for each entry

### ISSUE THREAD

**Issue by pawlisko80** [opnsense/plugins#3233]
**Important notices**
Before you add a new report, we ask you kindly to acknowledge the following:

- [X] I have read the contributing guide lines at https://github.com/opnsense/plugins/blob/master/CONTRIBUTING.md
- [X] I have searched the existing issues, open and closed, and I'm convinced that mine is new.
- [X] When the request is meant for an existing plugin, I've added its name to the title.

**Is your feature request related to a problem? Please describe.**
Ability to choose which address is being sent to DDNS service - IPv4/IPv6/both using entered credentials.
As expansion - the ability to choose which address within Interface to send if multiple addresses exist - i.e. VIPs
 
Within GUI I would use a drop-down menu (IPv4/IPv6/IPv4+IPv6), same with expansion (addres#1, address #2, VIP#1, VIP#2, etc).

**Comment by TheDom42** [comment]
I was looking for the same feature. I remember the discussion for the old DynDNS client on https://forum.opnsense.org/index.php?topic=16948.0 and https://github.com/opnsense/plugins/pull/2053 respectively.
I believe this is not really implemented (yet) with the new client.
A workaround (I personally have not yet tested, as I am using INWX as my provider who only supports updating v4 and v6 within the same call) used to be (still is?) to add two entries for the same provider and using a v4 for one and a v6 method for the other entry.

However, I would strongly support the request made above, as I would like to use the "Interface" method instead of relying on an external service to get my IP. Opnsense already knows v4 and v6 for my WAN, so why generate unnecessary traffic to servers on the net?

**Comment by tduboys** [comment]
Hi,
I looked on how ddclient works, and it have a ipv4/ipv6 logic (not implemented on every providers yet, but work in progress).
Opnsense plugin generate a `use` option, but this one is deprecated, and we need to use `usev4` and `usev6` option.

For example :
Current plugin creates this config file
```
syslog=yes                  # log update msgs to syslog
pid=/var/run/ddclient.pid   # record PID in file.
ssl=yes
verbose=yes

use=if, if=igb0_vlan832, \
protocol=dyndns2, \
server=dyndns.mydomain.com, \
login=user, \
password=xxxxx \
opnsense.mydomain.com
```

And it now should create
```
syslog=yes                  # log update msgs to syslog
pid=/var/run/ddclient.pid   # record PID in file.
ssl=yes
verbose=yes

usev4=ifv4, ifv4=igb0_vlan832, \
usev6=ifv6, ifv6=bridge0, \
protocol=dyndns2, \
server=dyndns.mydomain.com, \
login=user, \
password=xxxxx \
opnsense.mydomain.com
```
and it can be possible to define a different way to retrieve ipv4/ipv6 address.

ddclient already have a logic that, if provider dont support both ips (like dyndns2 provider, I just made a PR on ddclient for that), the old way still works.

We probably need to duplicate the checkip field on the web interface, to define a method for ipv4 and another for ipv6.

I dont know how to develop/test plugins for opnsense, but I can try to made a patch if someone can explain how to test it :)

**Comment by AdSchellevis** [comment]
When using the latest OPNsense edition, it might also be an option to take a look at our backend selection introduced in https://github.com/opnsense/plugins/commit/ef91a6b4f9feef7b412ab4a98f14f452d0c187ac , the new code is probably easier to extend (two examples are included).

**Comment by tduboys** [comment]
Hi,
I looked on the ddclient_opn backend, and it seems cant use both ipv4 and ipv6 address, for dyndns2 agent.
https://github.com/opnsense/plugins/blob/master/dns/ddclient/src/opnsense/scripts/ddclient/lib/account/dyndns2.py#L78
We also need to update this backend.

But for the ddclient part, I suggest to add ipv4 and ipv6 check methods like that
![image](https://user-images.githubusercontent.com/1507743/215346717-a5b20569-80f0-43b2-ab80-12c312548335.png)

and add a None section for both check. We coult generate a ddclient.conf file with the correct options, and handle more ddns scenarios (like ipv4 and ipv6 in different interfaces, etc…)

**Comment by AdSchellevis** [comment]
> I looked on the ddclient_opn backend, and it seems cant use both ipv4 and ipv6 address, for dyndns2 agent.

You need to register two services, one for ipv4 and one for ipv6 if you need dual stack. If I'm not mistaken most services require two separate calls to update two addresses anyway.

**Comment by tduboys** [comment]
On the way ddclient works, it merges configs that are for the same hostname.
It's why there are these bugs https://github.com/opnsense/plugins/issues/3249 or https://github.com/opnsense/plugins/issues/3149

So we need to know at least if a config is for ipv4 or ipv6, to generate the appropriate config.
The 2 ways I see is : 
- create 2 configs with the same creds, but one for ipv4 and one for ipv6, but we need to add a field to say if we look for a v4 or v6 address (that generates `usev4` or `usev6` ddclient config)
- create only one config with ipv4 and ipv6 config.

In both cases, ddclient will merge all the config in only one and passing it to the provider that will do one or two calls depending on the provider capacity.

**Comment by AdSchellevis** [comment]
I'll let you play with it, I mostly added an alternative method to simplify things as far as possible, wasn't planning on spending much time on it myself at the moment (maybe adding a couple of more examples if I have some accounts to test with at some point).

Eventually the old plugin will (os-dyndns) will go away on our end,  the we invest is only aimed at making sure that happens rather sooner than later.

**Comment by tduboys** [comment]
I think the simplier method for users is to add a IPv4/IPv6 selector on each config, to define what kind of IP we need to check.
It should work with ddclient and the local ddcliens_opn with two configs, and less confusing for every people.

I want to help to do that, but I cant find a simple way to test a local plugin. I need to read the docs… :)

**Comment by Starkstromkonsument** [comment]
> > I looked on the ddclient_opn backend, and it seems cant use both ipv4 and ipv6 address, for dyndns2 agent.
> 
> You need to register two services, one for ipv4 and one for ipv6 if you need dual stack. If I'm not mistaken most services require two separate calls to update two addresses anyway.

I don't understand why some services require to update IPv4 an IPv6 in one call either. But unfortunately they are out there (in my case: INWX).

I would be really happy if some would come up witch a fix for this. Due to my limited time I can only offer help with testing.

**Comment by Starkstromkonsument** [comment]
I did some more research: ddclient has a new maintainer and IPv6-Support was added in Version v3.11.0 (see: https://github.com/ddclient/ddclient/issues/172).

The new syntax of ddclient (using usev4/usev6 and ifv4/ifv6) is already supported by os-ddclient as of version 1.13. 

Unfortunately INWX requires a custom URL-format for IPv6 (`https://dyndns.inwx.com/nic/update?myip=<ipaddr>&myipv6=<ip6addr>`) which does not comply with the dyndns2-protocol (see: https://github.com/ddclient/ddclient/issues/96#issuecomment-2184201890). So there is nothing OPNsense can do about the INWX-issue for the ddclient backend.

My question is: will OPNsense keep using/supporting the ddclient backend or is the plan to deprecate it in favor of the native backend?

**Comment by Monviech** [comment]
@Starkstromkonsument The os-caddy plugin supports inwx since 1.5.7. You can run the plugin in dyndns only mode and you can choose between ipv4/ipv6 or both to be updated.

https://docs.opnsense.org/manual/how-tos/caddy.html#use-dynamic-dns-in-client-mode-only

Just saying this as an additional option.

**Comment by AdSchellevis** [comment]
> My question is: will OPNsense keep using/supporting the ddclient backend or is the plan to deprecate it in favor of the native backend?

Since it seems to be maintained again, we have no plans to deprecate it.

**Comment by Starkstromkonsument** [comment]
> So there is nothing OPNsense can do about the INWX-issue for the ddclient backend.

FYI: The fix https://github.com/ddclient/ddclient/pull/690 for INWX has been merged into ddclient today. It will be part oft the next release.

**Comment by Jannik-Hm** [comment]
Why is this still not supported when using the ddclient backend? (my provider doesn't exist in the native backend)
Since my opnsense is sitting behind another router, I am forced to use a web method for ipv4, but still need the interface method for ipv6. 

It's been 1.5 years since @tduboys provided a mockup.
Is it that hard to implement?

I'm open to coding myself, if one is providing me with a quick guide on what files to edit and how to test the changes. 
(never worked on plugins before)

Otherwise is there a simple way to manually edit the ddclient.conf file persistently as a workaround?

**Comment by Starkstromkonsument** [comment]
This is because ddclient has not published a new [release](https://github.com/ddclient/ddclient/releases) since Nov 23, 2023 (version 3.11.2). OPNsense provides the latest version (3.11.2) from [FreeBSD Ports](https://ports.freebsd.org/cgi/ports.cgi?query=ddclient&stype=all&sektion=all). Please refer to https://github.com/ddclient/ddclient/issues/552 for further information.

**Comment by madmurl0c** [comment]
Is it not possible to use mixed methods for getting the IP addresses? I tried it like mentioned in a previous comment.

```
usev6=ifv6, ifv6=..., \
usev4=cmd, cmd="...", \
```

It gives me this error: `WARNING: file /usr/local/etc/ddclient.conf, line 11: Invalid Value for keyword 'usev4' = 'cmd'`

I need to use the v6 address of the interface but the v4 address from the external check.

@Jannik-Hm Did you manage to solve/bypass this in any way?

**Comment by Jannik-Hm** [comment]
> @Jannik-Hm Did you manage to solve/bypass this in any way?
> 
> Edit: I managed to get rid of the error message like so:
> 
> usev6=ifv6, ifv6=..., \
> usev4=cmdv4, cmdv4="...", \

I finally switched over to creating a custom script. 

You can change ddclient to use different options manually in the cli, like you did, but after every reboot it gets reverted to what is configured in the opnsense UI.

Since I am using static ipv6 suffixes for all my public services I wanted a way to set it together with the specific domain and my dns updater will centrally update every prefix in the dns service.

I created a simple python script which handles this very well for my usecase.

**Comment by madmurl0c** [comment]
Thanks for your reply @Jannik-Hm but since I need it directly on my opnsense box I had to get creative. I worked around this issue by saving my custom config in a separate file `ddclient.conf.bak` and then creating a cron job that restores my custom config every 30min. It's not a clean solution but it'll have to do until its possible via the UI.

**Comment by Jannik-Hm** [comment]
It's not stupid if it works @madmurl0c 😂.

Knowing how many open issues and pr's regarding plugins exist, I guess it will never be implemented. 
Opened a request for a true/false toggle for a specific nginx feature, never got any response.

Plugins seem very neat but my feeling is that the moment you want anything more than basic, plugins won't allow you to do it. For that reason I had to move a few services away from OPNSense.

**Comment by pawlisko80** [comment]
A bit reviving that thread :)

Problem at hand - I need to update Dynamic IPs. My DynDNS provider - HE.net.

https://[dns.he.net/docs.html](https://dns.he.net/docs.html)

This is their documentation which is quite interesting from many points.
1. "If you have tagged both an A and AAAA record to be dynamic, you will see the icon twice, it is only necessary to generate one key as it is bound to the name of the record and not the name/type (...)" and here is the first issue. Both A and AAAA records MUST have the same key/password to work.
2. "When making updates, you will need to make a separate update for ipv4 and ipv6." and this is when all above good wishes are being dead.
HE expects 2 lines:
"https://dyn.dns.he.net/nic/update?hostname=dyn.example.com&password=password&myip=192.168.0.1"
"https://dyn.dns.he.net/nic/update?hostname=dyn.example.com&password=password&myip=2001:db8:beef:cafe::1"
But when I go for ddclient as backend change ddclient.conf to make it work instead of 2 calls there is only one : 
"https://dyn.dns.he.net/nic/update?hostname=dyn.example.com&password=password&myip=192.168.0.1,2001:db8:beef:cafe::1"which gets me: "RECEIVE: badip".

So even any changes added to ddclient.conf or others and not working as client with combine IPs in one update while I need to be 2.

So to summarize -I need 2 calls, but I have only 1.

With the original issue as stated, 2 different setups (for different versions of IP versions) could potentially resolve this issue.

Any comments/suggestions are welcomed.

**Comment by TheDom42** [comment]
Doesn't the current workaround of using two separate entries for v4 and v6 work? Of course, it would be nice to have it in one interface. But it should generally already work?

However, the major issue at hand was also the fact that some services (INWX for example) only accept updating both at the same time (because the second update deletes the first update even when they are not the same version - making a parallel update in one call necessary). This was (and still is not possible) until (at least for INWX) ddclient 4.0 is ported.

**Comment by pawlisko80** [comment]
OK, so let me be very clear.

I have 3 records I am updating. 1 x A and 2 x AAAA. Reason being access from the road for WireGuard server, and certain crazy IPv4/IPv6 implementation I found. Basically iOS WG client is prioritizing IPv4 over IPv6, which sounds good until you are on the IPv6 only network with IPv4 being handled as IPv6 over IPv4. Even though WG is UDP it acts crazy, the only resolution is separate connection with AAAA record without A one.

My ddclient.conf looks like this:
`syslog=yes                  # log update msgs to syslog
pid=/var/run/ddclient.pid   # record PID in file.
verbose=yes
ssl=yes

#use=cmd, cmd="/usr/local/opnsense/scripts/ddclient/checkip -i ix0 -t 1 -s he --timeout 10", \
usev4=ifv4, ifv4=ix0, \
protocol=dyndns2, \
server=dyn.dns.he.net, \
login=dyn.example.com, \
password=password1 \
dyn.example.com

use=cmd, cmd="/usr/local/opnsense/scripts/ddclient/checkip -i ix0 -t 1 -s he --timeout 10", \
protocol=dyndns2, \
server=dyn.dns.he.net, \
login=dyn-ipv6.example.com, \
password=password2 \
dyn-ipv6.example.com

#use=cmd, cmd="/usr/local/opnsense/scripts/ddclient/checkip -i ix0 -t 1 -s he --timeout 10", \
usev6=ifv6, ifv6=ix0, \
protocol=dyndns2,
server=dyn.dns.he.net, \
login=dyn.example.com, \
password=password1 \
dyn.example.com`

dyn-ipv6.example.com - updates perfectly
dyn.example.com - this is where I have as result:
'https://dyn.dns.he.net/nic/update?hostname=dyn.example.com&password=password1&myip=192.168.0.1,2001:db8:beef:cafe::1' which gets me: 'RECEIVE: badip'

I can make it work for either ipv4 only or ipv6 but not for both.
